lib/engine/tile.rb

```diff
@@ -90,7 +90,8 @@ def self.part(type, params, cache)
                               visit_cost: params['visit_cost'],
                               route: params['route'],
                               format: params['format'],
-                              loc: params['loc'])
+                              loc: params['loc'],
+                              node_id: cache.size)
```

Comment:
why do you need to add a node_id?

Reply:
Because I needed an id that was unique across all forms of nodes for use in disambiguating nodes in intra-tile paths/connections. Tile generates indexes per class, but those indexes aren't unique across the node classes (i.e. there can be a City with index 0 as well as a Town with index 0).

Reply:
I also needed node_id to be completely deterministic and based on the tile definition so that upgraded tiles will work correctly with existing routes/connections.

Reply:
let's get rid of node_id and just change the part.index = index to not group by class, would that work?

Reply:
Will the be guaranteed to result in the same index for a node after an upgrade? Since node_id is a function of the tile definition, as long as the node appears in the same order in both the original tile definition and it's upgraded tile, it will work correctly.

Reply:
It will work as long as all city and town definitions occur before any paths are defined. I can live with that.

Reply:
Done.